src/components/SearchBar.jsx

```diff
@@ -0,0 +1,187 @@
+        const serviceOrigin = url.split('/', 3).join('/')
+        
+        this.sources.push({
+          slug: service.slug,     // can be used to show where a suggestion comes from
+          origin: serviceOrigin,
+          id: intent._id,
+          ready: false,
+          window: null,           // will hold a reference to the window we're sending messages to
+          resolve: null           // a reference to a function to call when the source sends suggestions
+        })
+      })
+      
+      window.addEventListener('message', this.onMessageFromSource(this.sources))
+    })
+  }
+  
+  onMessageFromSource = (sources) => (event) => {
+    // this re-implements a subset of injectService found in lib/intents, though only the part that are useful for suggestions
+    const source = sources.find(source => source.origin === event.origin)
+    
+    if (!source) return
+    
+    if (event.data.type === `intent-${source.id}:ready`) {
+      source.ready = true
+      source.window = event.source
+      
+      source.window.postMessage({}, event.origin)
+    }
+    else if (event.data.type === `intent-${source.id}:data` && source.resolve) {
+      source.resolve({
+        id: source.id,
+        suggestions: event.data.suggestions
+      })
+      source.resolve = null
+    }
+    else {
+      console.log('unhandled message:', event)
+    }
+  }
+  
+  onChange = (event, { newValue }) => {
+    this.setState({
+      query: newValue
+    })
+  }
+  
+  onSuggestionsFetchRequested = ({ value }) => {
+    this.setState({
+      suggestions: []
+    })
+    
+    this.sources.filter(source => source.ready).forEach(async (source) => {
+      const {id, suggestions} = await new Promise(resolve => {
+        source.resolve = resolve
+        source.window.postMessage({ query: value }, source.origin)
+      })
+      
+      this.setState(state => ({
+        ...state,
+        suggestions: [
+          ...state.suggestions,
+          {
+            title: this.sources.find(source => source.id === id).slug,
+            suggestions
+          }
+        ]
+      }))
+    })
+  }
+  
+  onSuggestionsClearRequested = () => {
+    this.setState({
+      suggestions: []
+    })
+  }
+  
+  onSuggestionSelected = (event, { suggestion }) => {
+    const { onSelect } = suggestion
+    // `onSelect` is a string that describes what should happen when the suggestion is selected. Currently, the only format we're supporting is `open:http://example.com` to change the url of the current page.
+    
+    if (/^open:/.test(onSelect)) {
+      const url = onSelect.substr(5)
+      window.location.href = url
+    }
+    else {
+      console.log('suggestion onSelect (' + onSelect + ') could not be executed')
+    }
+    
+    this.setState({ query: '' })
+  }
+  
+  getSectionSuggestions = (section) => section.suggestions.slice(0, 10) // arbitrary limit, should probably depend on the number of sections
```

Comment:
Don't use unnamed magic number, or in other words, name your magic number to explicit your intention:

```js
const ARBITRARY_LIMIT_DEPENDS_ON_SECTIONS_NUMBER = 10
getSectionSuggestions = (section) => section.suggestions.slice(0, ARBITRARY_LIMIT_DEPENDS_ON_SECTIONS_NUMBER)
```

And magic: there is no need of comments anymore.
```